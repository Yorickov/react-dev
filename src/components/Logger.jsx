import _ from 'lodash';
import React, { Component } from 'react';

class Logger extends Component {
  state = { items: [] };

  handleCount = (value) => {
    const { items } = this.state;
    const currentValue = _.get(items, [0, 'value'], 0) + value;
    const current = { id: _.uniqueId(), value: currentValue };
    this.setState({ items: [current, ...items] });
  };

  // handleCount(value) {
  //   const { items } = this.state;
  //   const currentValue = _.get(items, [0, 'value'], 0) + value;
  //   const current = { id: _.uniqueId(), value: currentValue };
  //   this.setState({ items: [current, ...items] });
  // }

  handleInc = () => this.handleCount(1);

  handleDec = () => this.handleCount(-1);

  handleRemove = (currentId) => () => {
    const { items } = this.state;
    this.setState({ items: items.filter(({ id }) => id !== currentId) });
  };

  renderItem({ id, value }) {
    return (
      <button key={id} type="button" className="list-group-item" onClick={this.handleRemove(id)}>
        {value}
      </button>
    );
  }

  renderBtns() {
    return (
      <div className="btn-group" role="group">
        <button type="button" className="btn inc" onClick={this.handleInc}>+</button>
        <button type="button" className="btn dec" onClick={this.handleDec}>-</button>
      </div>
    );
  }

  renderLog() {
    const { items } = this.state;
    if (items.length === 0) {
      return null;
    }
    return (
      <div className="list-group">
        {items.map((item) => this.renderItem(item))}
      </div>
    );
  }

  render() {
    return (
      <div>
        <h5>Logger</h5>
        {this.renderBtns()}
        {this.renderLog()}
      </div>
    );
  }
}

export default Logger;
