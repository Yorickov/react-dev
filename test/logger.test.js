import React from 'react';
import Logger from '../src/components/Logger.jsx';

it('Logger', () => {
  const wrapper = mount(<Logger />);
  expect(wrapper.render()).toMatchSnapshot();

  const inc = wrapper.find('.inc');
  const dec = wrapper.find('.dec');
  expect(wrapper.find('.list-group-item')).toHaveLength(0);

  inc.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  inc.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  dec.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
  expect(wrapper.find('.list-group-item')).toHaveLength(3);

  const items = wrapper.find('.list-group-item-action');
  items.at(1).simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
  expect(wrapper.find('.list-group-item')).toHaveLength(2);
});
