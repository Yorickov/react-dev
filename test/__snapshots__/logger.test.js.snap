// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger 1`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn dec"
      type="button"
    >
      -
    </button>
  </div>
</div>
`;

exports[`Logger 2`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`Logger 3`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`Logger 4`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`Logger 5`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;
